tools/test/travis-ci/scancode-evaluate.py

```diff
@@ -57,57 +92,64 @@ def license_check(directory_name, file):
 
     offenders = []
     try:
-        # find all licenses in the files, must be licensed and permissive
-        with open(file, 'r') as scancode_output:
-            results = json.load(scancode_output)
-    except ValueError:
-        userlog.warning("JSON could not be decoded")
+        with open(scancode_output, 'r') as read_file:
+            scancode_output_data = json.load(read_file)
+    except json.JSONDecodeError as jex:
+        userlog.warning("JSON could not be decoded, Invalid JSON in body: %s", jex)
         return -1
 
-    try:
-        for file in results['files']:
-            license_offender = {}
-            license_offender['file'] = file
-            # ignore directory, not relevant here
-            if license_offender['file']['type'] == 'directory':
+    if 'files' not in scancode_output_data:
+        userlog.warning("Missing `files` attribute in %s" % (scancode_output))
+        return -1
+
+    for scancode_output_data_file in scancode_output_data['files']:
+        try:
+            if scancode_output_data_file['type'] != 'file':
                 continue
-            if not license_offender['file']['licenses']:
-                license_offender['reason'] = MISSING_LICENSE_TEXT
-                offenders.append(license_offender)
+        except KeyError as e:
+            userlog.warning("Could not find %s attribute in %s" % (str(e), scancode_output))
+            return -1
+
+        try:
+            if not scancode_output_data_file['licenses']:
+                scancode_output_data_file['fail_reason'] = MISSING_LICENSE_TEXT
+                offenders.append(scancode_output_data_file)
+                # check the next file in the scancode output
                 continue
-
-            found_spdx = False
-            for i in range(len(license_offender['file']['licenses'])):
-                if license_offender['file']['licenses'][i]['category'] != 'Permissive':
-                    license_offender['reason'] = MISSING_PERMISIVE_LICENSE_TEXT
-                    offenders.append(license_offender)
-                # find SPDX, it shall be one of licenses found
-                if license_offender['file']['licenses'][i]['matched_rule']['identifier'].find("spdx") != -1:
-                    found_spdx = True
-
-            if not found_spdx:
+        except KeyError as e:
+            userlog.warning("Could not find %s attribute in %s" % (str(e), scancode_output))
+            return -1
+
+        try:
+            if not has_permissive_text_in_scancode_output(scancode_output_data_file['licenses']):
+                scancode_output_data_file['fail_reason'] = MISSING_PERMISIVE_LICENSE_TEXT
+                offenders.append(scancode_output_data_file)
+        except KeyError as e:
```

Comment:
Which `KeyError` are we trying to trap here? `...['licenses']` or `...['fail_reason']`? Is it actually possible these keys don't exist at this point? Do we really need to handle this `KeyError`?

Reply:
There was a larger `try` section to catch all `KeyError` exceptions. I instead created multiple `try` sections to cover all possibilities as they were before, just in case it was needed. The idea was to have the same behaviour except for what needed to be changed to fix the bug whilst improving maintainability a tiny bit.